### ISSUE TITLE
debug_assert_with_mut_call: false positive on immutable method call

### ISSUE THREAD

**Issue by taiki-e** [rust-lang/rust-clippy#4737]
Code:
```rust
pub fn f() {
    let v = &mut &mut Vec::new();
    debug_assert!(v.is_empty()); //~ ERROR do not call a function with mutable arguments inside of `debug_assert!`
    v.push(());
}
```

Error:
```text
error: do not call a function with mutable arguments inside of `debug_assert!`
  --> rust/src/lib.rs:17:19
   |
17 |     debug_assert!(v.is_empty());
   |                   ^^^^^^^^^^^^
   |
   = note: `#[deny(clippy::debug_assert_with_mut_call)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#debug_assert_with_mut_call
```

**Comment by flip1995** [comment]
A workaround for this is to deref the vector once:
```rust
debug_assert!((*v).is_empty());
```

**Comment by philipc** [comment]
Similar issue for code:
```
use core::cell::{Ref, RefCell};

pub fn f() {
    let mut x = (true,);
    let cell = RefCell::new((&mut x,));
    let r = Ref::map(cell.borrow(), |x| &x.0);
    debug_assert!(r.0);
}
```
Error:
```
error: do not call a function with mutable arguments inside of `debug_assert!`
 --> src/lib.rs:7:19
  |
7 |     debug_assert!(r.0);
  |                   ^^^
  |
  = note: `#[deny(clippy::debug_assert_with_mut_call)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#debug_assert_with_mut_call
```
